# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [1]:
(lambda num: num + 2)(5)

7

#### Saving to a Variable

In [2]:
add_nums = lambda num1, num2: num1 + num2

add_nums(4,5)

9

#### Multiple Inputs

In [4]:
(lambda string1, string2: f'{string1} {string2}')('foo','bar')

'foo bar'

#### Passing a Lambda into a Function

In [6]:
def math_operations(func, num):
    return func(num)


math_operations(lambda num: num * 2, 8)

math_operations(lambda num: num // 2, 50)

25

#### Returning a Lambda from a Function

In [12]:
def outter_func(str1):
    inner_func_var = 'foo'
    
    
    return lambda astring: f'{inner_func_var} {str1}'

addToFooBar = outter_func('bar')

addToFooBar('spam')

'foo bar'

#### If Statements within Lambdas

In [16]:
def foobar():
    if True:
        return True
    else:
        return False
    
upperVowel = lambda astring: astring.upper() if astring[0].lower() in 'aeiou' else astring

upperVowel('apple')


'APPLE'

#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [17]:


f = lambda num: num ** 3 


print(f(5))
print(f(25))
print(f(30))

125
15625
27000


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [22]:
print(map(lambda astring: astring.upper(), ['foo', 'bar', 'spam', 'eggs']))

print(list(map(lambda astring: astring.upper(), ['foo', 'bar', 'spam', 'eggs'])))


def capitalize_string(astring):
    return astring.upper()

list(map(capitalize_string, ['foo', 'bar']))



[word for word in map(capitalize_string, ['foo', 'bar'])]

['FOO', 'BAR', 'SPAM', 'EGGS']


['FOO', 'BAR']

#### Using Lambda's with Map

In [23]:
nums_plus_ten = map( lambda num: num + 10, [3,4,5,6,12])

list(nums_plus_ten)

[13, 14, 15, 16, 22]

#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [24]:
nums_times_minus = map(lambda num: num * 2 - 1, [1,2,3,4,5,6,7,8,9,10])

list(nums_times_minus)

[1, 3, 5, 7, 9, 11, 13, 15, 17, 19]

## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [25]:
list(filter(lambda astring: True, [12,3,1,3,18]))

[12, 3, 1, 3, 18]

#### Using Lambda's with Filter()

In [31]:
list(filter(lambda num: True if num % 2 == 0 else False, [3,2,6,4,5,9,1]))


students = ['robin', 'tony', "rhett", 'zack', 'zoey']

[word for word in filter(lambda name: name[0].lower() != 'z', students)]



def divisible_by_ten(num):
    return num % 10 == 0


[num for num in filter(divisible_by_ten, [10,33,20,30,455])]

[10, 20, 30]

#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [34]:
from statistics import mean

my_list = [23, 57, 42, 100, 37]

def my_mean(num):
    return num <= mean(my_list)

[num for num in filter(my_mean, my_list)]
    
    

[23, 42, 37]

In [37]:
astring = 'foo bar fizz'
letters = []

for letter in astring:
    letters.append(letter)
print(letters)

[letter for letter in astring]

['f', 'o', 'o', ' ', 'b', 'a', 'r', ' ', 'f', 'i', 'z', 'z']


['f', 'o', 'o', ' ', 'b', 'a', 'r', ' ', 'f', 'i', 'z', 'z']

## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [45]:
from functools import reduce

sum([1,2,3,4])

reduce(lambda agg, num: agg + num, [1,2,3,4,5,6,7,8,9])

45

#### Using Lambda's with Reduce()

In [ ]:
#See Above

#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [42]:
reduce(lambda agg, num: agg * num, [1,2,3,4,5])

120

## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [48]:
def add_recursive(num):
    if num == 0:
        return num
    print(f"add_recursive({num}) = {num} + add_recursive({num - 1})")
    return add_recursive(num - 1) + num


add_recursive(5)

add_recursive(5) = 5 + add_recursive(4)
add_recursive(4) = 4 + add_recursive(3)
add_recursive(3) = 3 + add_recursive(2)
add_recursive(2) = 2 + add_recursive(1)
add_recursive(1) = 1 + add_recursive(0)


15

#### Writing a Factorial Function

In [49]:
def find_factorial(num):
    if num ==1:
        print(f'find_factorial({num}) = {num}')
        return num
    print(f'find_factorial({num}) = {num} * find_factorial({num - 1})')
    return num * find_factorial(num - 1)

find_factorial(5)

find_factorial(5) = 5 * find_factorial(4)
find_factorial(4) = 4 * find_factorial(3)
find_factorial(3) = 3 * find_factorial(2)
find_factorial(2) = 2 * find_factorial(1)
find_factorial(1) = 1


120

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [50]:
def subtract(num):
    if num == 1:
        return num
    
    return num - subtract(num-1)


subtract(5)

3

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [ ]:
def ranger(start,stop,step):
    current = start
    while start < stop:
        yield current
        current += step
        
        
my_range = ranger(1,10,2)

print(my_range)

print([num for num in my_range])

<generator object ranger at 0x107d68ac0>


#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [8]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

[word for word in filter(lambda city: city.strip() != "", places)]

['Argentina', 'San Diego', 'Boston', 'New York']

### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [19]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

sortedNames = sorted(author, key = lambda name: name.split()[-1])
print(sortedNames)

['Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'Victor aNisimov', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [10]:
# F = (9/5)*C + 32
# nums_times_minus = map(lambda num: num * 2 - 1, [1,2,3,4,5,6,7,8,9,10])
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]


temp_in_f = list(map(lambda x: (x[0], x[1] * 9/5 + 32), places)) 

print(temp_in_f)

[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [34]:
def fib(num):
    if num <= 0:
        print(f'fib({num}) = {num}')
        return num
    elif num == 1:
        print(f'fib({num}) = {num}')
        
        return num
#                                     This one is throwing me for a loop. not sure what i am doing wrong
    else:
        new_num = fib(num - 1) + fib(num - 2)
        print(f'fib{num} = fib({num} - 1) + fib({num} - 2)')
        return new_num
    
fib(5)

fib(1) = 1
fib(0) = 0
fib2 = fib(2 - 1) + fib(2 - 2)
fib(1) = 1
fib3 = fib(3 - 1) + fib(3 - 2)
fib(1) = 1
fib(0) = 0
fib2 = fib(2 - 1) + fib(2 - 2)
fib4 = fib(4 - 1) + fib(4 - 2)
fib(1) = 1
fib(0) = 0
fib2 = fib(2 - 1) + fib(2 - 2)
fib(1) = 1
fib3 = fib(3 - 1) + fib(3 - 2)
fib5 = fib(5 - 1) + fib(5 - 2)


5